textutils-2.0/man/split.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH SPLIT "1" "July 2001" "GNU textutils 2.0" FSF
.SH NAME
split \- split a file into pieces
.SH SYNOPSIS
.B split
[\fIOPTION\fR] [\fIINPUT \fR[\fIPREFIX\fR]]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Output fixed-size pieces of INPUT to PREFIXaa, PREFIXab, ...; default
PREFIX is `x'.  With no INPUT, or when INPUT is -, read standard input.
.TP
\fB\-b\fR, \fB\-\-bytes\fR=\fISIZE\fR
put SIZE bytes per output file
.TP
\fB\-C\fR, \fB\-\-line\-bytes\fR=\fISIZE\fR
put at most SIZE bytes of lines per output file
.TP
\fB\-l\fR, \fB\-\-lines\fR=\fINUMBER\fR
put NUMBER lines per output file
.TP
\fB\-NUMBER\fR
same as \fB\-l\fR NUMBER
.TP
\fB\-\-verbose\fR
print a diagnostic to standard error just
before each output file is opened
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
SIZE may have a multiplier suffix: b for 512, k for 1K, m for 1 Meg.
.SH AUTHOR
Written by Torbjorn Granlund and Richard M. Stallman.
.SH "REPORTING BUGS"
Report bugs to <bug-textutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 1999 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B split
is maintained as a Texinfo manual.  If the
.B info
and
.B split
programs are properly installed at your site, the command
.IP
.B info split
.PP
should give you access to the complete manual.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



mawk-1.3.3/split.c

/********************************************
split.c
copyright 1991, Michael D. Brennan

This is a source file for mawk, an implementation of
the AWK programming language.

Mawk is distributed without warranty under the terms of
the GNU General Public License, version 2, 1991.
********************************************/

/* $Log: split.c,v $
 * Revision 1.3  1996/02/01  04:39:42  mike
 * dynamic array scheme
 *
 * Revision 1.2  1993/07/15  01:55:03  mike
 * rm SIZE_T & indent
 *
 * Revision 1.1.1.1  1993/07/03	 18:58:21  mike
 * move source to cvs
 *
 * Revision 5.4	 1993/05/08  18:06:00  mike
 * null_split
 *
 * Revision 5.3	 1993/01/01  21:30:48  mike
 * split new_STRING() into new_STRING and new_STRING0
 *
 * Revision 5.2	 1992/07/08  21:19:09  brennan
 * patch2
 * change in split() requires that
 * bi_split() call load_array() even
 * when cnt is 0.
 *
 * Revision 5.1	 1991/12/05  07:56:31  brennan
 * 1.1 pre-release
 *
*/

/* split.c */


/* For all splitting up to MAX_SPLIT fields go into
   split_buff[], the rest go onto split_ov_list ( split
   overflow list)

   We can split one of three ways:
     (1) By space:
	 space_split() and space_ov_split()
     (2) By regular expression:
	 re_split()    and re_ov_split()
     (3) By "" (null -- split into characters)
	 null_split() and null_ov_split()
*/

#define	 TEMPBUFF_GOES_HERE

#include "mawk.h"
#include "symtype.h"
#include "bi_vars.h"
#include "bi_funct.h"
#include "memory.h"
#include "scan.h"
#include "regexp.h"
#include "field.h"

SPLIT_OV *split_ov_list ;

static int PROTO(re_ov_split, (char *, PTR)) ;
static int PROTO(space_ov_split, (char *, char *)) ;
static int PROTO(null_ov_split, (char *)) ;

/* split string s of length slen on SPACE without changing s.
   load the pieces into STRINGS and ptrs into
   split_buff[]
   return the number of pieces */

int
space_split(s, slen)
   register char *s ;
   unsigned slen ;
{
   char *back = s + slen ;
   int i = 0 ;
   int len ;
   char *q ;
   STRING *sval ;
   int lcnt = MAX_SPLIT / 3 ;

#define EAT_SPACE()   while ( scan_code[*(unsigned char*)s] ==\
			      SC_SPACE )  s++
#define EAT_NON_SPACE()	  \
    *back = ' ' ; /* sentinel */\
    while ( scan_code[*(unsigned char*)s] != SC_SPACE )	 s++ ;\
    *back = 0


   while (lcnt--)
   {
      EAT_SPACE() ;
      if (*s == 0)  goto done ;
      /* mark the front with q */
      q = s++ ;
      EAT_NON_SPACE() ;
      sval = split_buff[i++] = new_STRING0(len = s - q) ;
      memcpy(sval->str, q, len) ;

      EAT_SPACE() ;
      if (*s == 0)  goto done ;
      q = s++ ;
      EAT_NON_SPACE() ;
      sval = split_buff[i++] = new_STRING0(len = s - q) ;
      memcpy(sval->str, q, len) ;

      EAT_SPACE() ;
      if (*s == 0)  goto done ;
      q = s++ ;
      EAT_NON_SPACE() ;
      sval = split_buff[i++] = new_STRING0(len = s - q) ;
      memcpy(sval->str, q, len) ;

   }
   /* we've overflowed */
   return i + space_ov_split(s, back) ;

 done:
   return i ;
}

static int
space_ov_split(s, back)
   register char *s ;
   char *back ;

{
   SPLIT_OV dummy ;
   register SPLIT_OV *tail = &dummy ;
   char *q ;
   int cnt = 0 ;
   unsigned len ;

   while (1)
   {
      EAT_SPACE() ;
      if (*s == 0)  break ;		    /* done */
      q = s++ ;
      EAT_NON_SPACE() ;

      tail = tail->link = ZMALLOC(SPLIT_OV) ;
      tail->sval = new_STRING0(len = s - q) ;
      memcpy(tail->sval->str, q, len) ;
      cnt++ ;
   }

   tail->link = (SPLIT_OV *) 0 ;
   split_ov_list = dummy.link ;
   return cnt ;
}

/* match a string with a regular expression, but
   only matches of positive length count */
char *
re_pos_match(s, re, lenp)
   register char *s ;
PTR re ; unsigned *lenp ;
{
   while ((s = REmatch(s, re, lenp)))
      if (*lenp)  return s ;
      else if (*s == 0)	 break ;
      else  s++ ;

   return (char *) 0 ;
}

int
re_split(s, re)
   char *s ;
   PTR re ;
{
   register char *t ;
   int i = 0 ;
   unsigned mlen, len ;
   STRING *sval ;
   int lcnt = MAX_SPLIT / 3 ;

   while (lcnt--)
   {
      if (!(t = re_pos_match(s, re, &mlen)))  goto done ;
      sval = split_buff[i++] = new_STRING0(len = t - s) ;
      memcpy(sval->str, s, len) ;
      s = t + mlen ;

      if (!(t = re_pos_match(s, re, &mlen)))  goto done ;
      sval = split_buff[i++] = new_STRING0(len = t - s) ;
      memcpy(sval->str, s, len) ;
      s = t + mlen ;

      if (!(t = re_pos_match(s, re, &mlen)))  goto done ;
      sval = split_buff[i++] = new_STRING0(len = t - s) ;
      memcpy(sval->str, s, len) ;
      s = t + mlen ;
   }
   /* we've overflowed */
   return i + re_ov_split(s, re) ;

done:
   split_buff[i++] = new_STRING(s) ;
   return i ;
}

/*
  we've overflowed split_buff[] , put
  the rest on the split_ov_list
  return number of pieces
*/

static int
re_ov_split(s, re)
   char *s ;
   PTR re ;
{
   SPLIT_OV dummy ;
   register SPLIT_OV *tail = &dummy ;
   int cnt = 1 ;
   char *t ;
   unsigned len, mlen ;

   while ((t = re_pos_match(s, re, &mlen)))
   {
      tail = tail->link = ZMALLOC(SPLIT_OV) ;
      tail->sval = new_STRING0(len = t - s) ;
      memcpy(tail->sval->str, s, len) ;
      s = t + mlen ;
      cnt++ ;
   }
   /* and one more */
   tail = tail->link = ZMALLOC(SPLIT_OV) ;
   tail->sval = new_STRING(s) ;
   tail->link = (SPLIT_OV *) 0 ;
   split_ov_list = dummy.link ;

   return cnt ;
}


int
null_split(s)
   char *s ;
{
   int cnt = 0 ;		 /* number of fields split */
   STRING *sval ;
   int i = 0 ;			 /* indexes split_buff[] */

   while (*s)
   {
      if (cnt == MAX_SPLIT)  return cnt + null_ov_split(s) ;

      sval = new_STRING0(1) ;
      sval->str[0] = *s++ ;
      split_buff[i++] = sval ;
      cnt++ ;
   }
   return cnt ;
}

static int
null_ov_split(s)
   char *s ;
{
   SPLIT_OV dummy ;
   SPLIT_OV *ovp = &dummy ;
   int cnt = 0 ;

   while (*s)
   {
      ovp = ovp->link = ZMALLOC(SPLIT_OV) ;
      ovp->sval = new_STRING0(1) ;
      ovp->sval->str[0] = *s++ ;
      cnt++ ;
   }
   ovp->link = (SPLIT_OV *) 0 ;
   split_ov_list = dummy.link ;
   return cnt ;
}


/*  split(s, X, r)
    split s into array X on r

    entry: sp[0] holds r
	   sp[-1] pts at X
	   sp[-2] holds s
*/
CELL *
bi_split(sp)
   register CELL *sp ;
{
   int cnt ;			 /* the number of pieces */


   if (sp->type < C_RE)	 cast_for_split(sp) ;
   /* can be C_RE, C_SPACE or C_SNULL */
   sp -= 2 ;
   if (sp->type < C_STRING)  cast1_to_s(sp) ;

   if (string(sp)->len == 0)	/* nothing to split */
      cnt = 0 ;
   else
      switch ((sp + 2)->type)
      {
	 case C_RE:
	    cnt = re_split(string(sp)->str, (sp + 2)->ptr) ;
	    break ;

	 case C_SPACE:
	    cnt = space_split(string(sp)->str, string(sp)->len) ;
	    break ;

	 case C_SNULL:		/* split on empty string */
	    cnt = null_split(string(sp)->str) ;
	    break ;

	 default:
	    bozo("bad splitting cell in bi_split") ;
      }


   free_STRING(string(sp)) ;
   sp->type = C_DOUBLE ;
   sp->dval = (double) cnt ;

   array_load((ARRAY) (sp + 1)->ptr, cnt) ;

   return sp ;
}
